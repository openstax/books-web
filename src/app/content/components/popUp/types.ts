import { LinkedArchiveTree, LinkedArchiveTreeNode, LinkedArchiveTreeSection } from '../../types';

export type LocationFilters = Map<
  string,
  { section: LinkedArchiveTreeNode, children?: LinkedArchiveTreeSection[] }
>;

export type LocationFiltersWithChildren = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;

export interface FiltersChange<T> {
  remove: T[];
  new: T[];
}
