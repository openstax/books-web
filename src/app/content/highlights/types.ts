import { Highlight, HighlightsSummary } from '@openstax/highlighter/dist/api';

export type HighlightData = Highlight;

export type CountsPerSource = Exclude<HighlightsSummary['countsPerSource'], undefined>;

export interface State {
  myHighlightsOpen: boolean;
  enabled: boolean;
  focused?: string;
  highlights: null | HighlightData[];
  summary: {
    filters: {
      colors: string[];
      chapters: string[];
    },
    pagination: {
      // even though we're manually splitting our requests out into smaller batches of sources,
      // if a source has more than [numPage] highlights the api will still paginate. in order to
      // access the subsequent pages, we need to store the original sources here, and continue to
      // use them until we run out of pages, then grab the next batch of sources. on the LAST
      // page the response size may be smaller than expected, in this case the next batch of
      // sources should be immediately queried to fill the remaining [numPage]. use
      // paginationUtils.getNextPageSources to get the next batch of sources
      sources: string[];
      page: number;
    } | null,
    loading: boolean;
    // totalCounts should reflect the UNFILTERED state of user's data, used in representing
    // available filter options. this should be updated on create/delete and after auth change
    totalCounts: {[key: string]: number};
    // filteredTotalCounts should reflect the FILTERED state of user's data, used in calculating
    // pagination requests. this should be updated after filter change before requesting
    // filtered highlights (because you need this to query them)
    filteredTotalCounts: {[key: string]: number};
    highlights: {
      [chapterId: string]: {[pageId: string]: HighlightData[]}
    };
  };
}
