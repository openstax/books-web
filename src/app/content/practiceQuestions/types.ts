import { LinkedArchiveTree, LinkedArchiveTreeSection } from '../types';

export interface State {
  isEnabled: boolean;
  summary: PracticeQuestionsSummary | null;
  open: boolean;
  selectedSection: LinkedArchiveTreeSection | null;
  currentQuestionIndex: number | null;
  questions: PracticeQuestion[];
}

export interface PracticeQuestionsSummary {
  countsPerSource: {
    [key: string]: number;
  };
}

interface PracticeAnswer {
  id: number;
  content_html: string;
  correctness: '0.0' | '1.0';
  feedback_html: string;
}

export interface PracticeQuestion {
  group_uuid: string;
  uid: string;
  stem_html: string;
  answers: PracticeAnswer[];
}

export type PracticeQuestions = PracticeQuestion[];

export type PracticeQuestionsLocationFilters = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;
