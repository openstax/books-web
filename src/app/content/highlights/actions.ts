import { createStandardAction } from 'typesafe-actions';

export const enableHighlighting = createStandardAction('FeatureFlag/enableHighlighting')<void>();
