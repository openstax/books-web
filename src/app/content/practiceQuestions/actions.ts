import { createStandardAction } from 'typesafe-actions';
import { PracticeQuestionsSummary } from './types';

export const receivePracticeQuestionsSummary = createStandardAction(
  'Content/PracticeQuestions/Summary/receive'
)<PracticeQuestionsSummary>();
