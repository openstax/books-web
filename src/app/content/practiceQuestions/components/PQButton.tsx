import styled from 'styled-components/macro';
import Button from '../../../components/Button';
import { linkColor } from '../../../components/Typography';

// tslint:disable-next-line: variable-name
const PQButton = styled(Button)`
  ${(props: { disabled: boolean }) => props.disabled && `
    background-color: #f1f1f1;
    color: #c1c1c1;
    cursor: not-allowed;
    &:hover {
      background-color: #f1f1f1;
    }
  `}
  ${(props: { withoutBg?: boolean }) => props.withoutBg && `
    border: none;
    background-color: transparent;
    color: ${linkColor};
    font-weight: normal;
  `}
`;

export default PQButton;
