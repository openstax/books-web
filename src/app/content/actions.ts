import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as searchActions from './search/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();

export const closeCallToActionPopup = createStandardAction('Content/closeCallToActionPopup')();
export const openCallToActionPopup = createStandardAction('Content/openCallToActionPopup')();
