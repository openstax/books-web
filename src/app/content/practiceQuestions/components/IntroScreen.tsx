import React from 'react';
import { FormattedMessage } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import styled from 'styled-components/macro';
import { textRegularSize } from '../../../components/Typography';
import { nextQuestion } from '../actions';
import * as pqSelectors from '../selectors';
import PQButton from './PQButton';

// tslint:disable-next-line: variable-name
const IntroScreenWrapper = styled.div`
  text-align: center;

  ${PQButton} {
    display: block;
    margin: 2rem auto;
  }
`;

// tslint:disable-next-line: variable-name
const IntroScreenMessage = styled.span`
  ${textRegularSize}
  margin: 4rem 0;
`;

// tslint:disable-next-line: variable-name
const IntroScreen = () => {
  const questionsCount = useSelector(pqSelectors.questionsCount);
  const dispatch = useDispatch();

  return <IntroScreenWrapper>
    <IntroScreenMessage>
      <FormattedMessage id='i18n:practice-questions:popup:intro:message' values={{ questions: questionsCount }}>
        {(msg: string) => msg}
      </FormattedMessage>
    </IntroScreenMessage>
    <PQButton
      variant='primary'
      size='large'
      onClick={() => dispatch(nextQuestion())}
      data-analytics-label='Start now'
    >
      <FormattedMessage id='i18n:practice-questions:popup:intro:start'>
        {(msg: string) => msg}
      </FormattedMessage>
    </PQButton>
  </IntroScreenWrapper>;
};

export default IntroScreen;
