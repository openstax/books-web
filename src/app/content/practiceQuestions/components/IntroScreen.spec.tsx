import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import * as Services from '../../../context/Services';
import MessageProvider from '../../../MessageProvider';
import { Store } from '../../../types';
import { nextQuestion, setQuestions } from '../actions';
import { PracticeQuestions } from '../types';
import IntroScreen from './IntroScreen';
import PQButton from './PQButton';

describe('IntroScreen for practice questions', () => {
  let store: Store;
  let dispatch: jest.SpyInstance;
  let services: ReturnType<typeof createTestServices>;
  let render: () => JSX.Element;

  beforeEach(() => {
    store = createTestStore();
    dispatch = jest.spyOn(store, 'dispatch');
    services = createTestServices();
    render = () => <Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <IntroScreen />
        </MessageProvider>
      </Services.Provider>
    </Provider>;
  });

  it('renders properly for 1 question', () => {
    store.dispatch(setQuestions([{}, {}] as PracticeQuestions));
    const component = renderer.create(render());
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders properly for 2 questions', () => {
    store.dispatch(setQuestions([{}, {}] as PracticeQuestions));
    const component = renderer.create(render());
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('dispatch action onClick', () => {
    const component = renderer.create(render());

    const button = component.root.findByType(PQButton);

    renderer.act(() => {
      button.props.onClick();
    });

    expect(dispatch).toHaveBeenCalledWith(nextQuestion());
  });
});
