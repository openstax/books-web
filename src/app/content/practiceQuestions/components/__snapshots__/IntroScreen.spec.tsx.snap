// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntroScreen renders properly for 1 question 1`] = `
<div
  className="IntroScreen__IntroScreenWrapper-h8gm9y-0 bsGjos"
>
  <span
    className="IntroScreen__IntroScreenMessage-h8gm9y-1 hubyCw"
  >
    2 questions in this section
  </span>
  <button
    className="Button__PlainButton-ayg7nk-2 PQButton-ypamof-0 dWFayQ"
    data-analytics-label="Start now"
    onClick={[Function]}
  >
    Start now
  </button>
</div>
`;

exports[`IntroScreen renders properly for 2 questions 1`] = `
<div
  className="IntroScreen__IntroScreenWrapper-h8gm9y-0 bsGjos"
>
  <span
    className="IntroScreen__IntroScreenMessage-h8gm9y-1 hubyCw"
  >
    2 questions in this section
  </span>
  <button
    className="Button__PlainButton-ayg7nk-2 PQButton-ypamof-0 dWFayQ"
    data-analytics-label="Start now"
    onClick={[Function]}
  >
    Start now
  </button>
</div>
`;
